File: src/main/java/xyz/gucciclient/event/EventManager.java
package xyz.gucciclient.event;


// import me.enterman.materialClient.module.management.Mod;

import xyz.gucciclient.modules.Module;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.*;
import java.util.function.Predicate;

@SuppressWarnings("unchecked")
public class EventManager<T,O> {
	Predicate<O> postPredicate;
	public EventManager(Predicate<O> postPredicate){
		this.postPredicate = postPredicate;
	}
	public EventManager(){
		this.postPredicate = (o)->true;
	}
	private Map<Class<? extends T>, Map<Method,O>> objs = new HashMap<>();
	private static EventManager<Event, Module> instance = new EventManager<>((mod -> mod.getState()));
	private static EventManager<Event,Object> other = new EventManager<>((o) -> true);
	public void flushEvents(){
		objs.clear();
	}
	public static EventManager<Event,Object> getOtherEventManager(){
		return other;
	}
	public void register(O obj){
		for (Method method : obj.getClass().getDeclaredMethods()) {
			if (method.isAnnotationPresent(Sub.class)){

				if(method.getParameterCount() == 1){

					if (Event.class.isAssignableFrom(method.getParameterTypes()[0])){

						if (!method.isAccessible()){
							method.setAccessible(true);
						}
						Map<Method,O> l;
						if((l = objs.getOrDefault(method.getParameterTypes()[0],null)) != null){
							l.put(method,obj);
						} else {
							l = new HashMap<>();
							l.put(method,obj);
							objs.put((Class<? extends T>) method.getParameterTypes()[0],l);
						}
					}
				}
			}
		}
	}

	public static EventManager<Event, Module> getInstance() {
		return instance;
	}

	public static <P extends Event> P postAll(P event){

		// FIXED: 2020/2/8 Events not working as expected
		// FIXME: 2020/3/14 Events not working as expected
		instance.post(event);
		other.post(event);
		return event;
	}
	public void post(Event event){
		Map<Method,O> l;
		if((l = objs.getOrDefault(event.getClass(),null)) != null){
			l.forEach((method,obj) -> {
				if(postPredicate.test(obj)){
					try {
						method.invoke(obj,event);
					} catch (IllegalAccessException | InvocationTargetException e) {
						e.printStackTrace();
					}
				}
			});
		}
	}

}

File: src/main/java/xyz/gucciclient/event/EventManager.kt
package me.delphi.events

import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.Multimap
import me.delphi.Delphi


import me.delphi.Manager
import me.delphi.mod.Mod
import java.lang.invoke.MethodHandle
import java.lang.invoke.MethodHandles


class MethodInfo(val method:MethodHandle,val obj: Any){
    fun invoke(param:Any){
        val a : Any = method.invoke(obj,param)
    }
}
@Suppress("UNCHECKED_CAST")
class EventManager:Manager{
    val predicate:(Any) -> Boolean = {
        if (it is Mod)
            it.enabled || it.keepReg
        true
    }
    private var subscribers: Multimap<Class<out Event?>, MethodInfo> = ArrayListMultimap.create()
    fun checkForSubscribers(obj:Any){
        if(obj.javaClass.isAnnotationPresent(SubscriberClass::class.java))
            obj.javaClass.declaredMethods.forEach {
                if(it.parameterCount == 1)
                    if(Event::class.java.isAssignableFrom(it.parameterTypes[0])){
                        if(!it.isAccessible)
                            it.isAccessible = true
                        val l: MethodHandles.Lookup = MethodHandles.lookup()
                        subscribers.put(it.parameterTypes[0] as Class<out Event?>?, MethodInfo(l.unreflect(it),obj))
                    }
            }
    }
    fun <E:Event> post(event:E){
        subscribers.get(event.javaClass).forEach {
            if(predicate(it.obj))
                it.invoke(event)
        }
    }

    /**
     * Flushes the Manager.
     * Used for dynamic reloading.
     * */
    override fun flush() {
        subscribers.clear()
    }

    override fun initiate() {

    }
}

